### PR TITLE
Autogenerate typings

### PR DESCRIPTION
**Pull request by dimitrisnl**
Connects-to: #792
Change-type: minor
Signed-off-by: Dimitrios Lytras <dimitrios@balena.io>

### CODE REVIEW

**Comment by dimitrisnl** [comment]
@LucianBuzzo 	I'm not really sure how to proceed with the auto-generation. Any pointers?

**Comment by LucianBuzzo** [comment]
@dimitrisnl What issues are you facing?

**Comment by dimitrisnl** [comment]
@LucianBuzzo 

[Reading the documentation](https://www.typescriptlang.org/docs/handbook/compiler-options.html), I would expect that having the `--declaration` flag  while running `npm run build (rimraf dist && tsc && npm run build:docs)` a `.d.ts` file would be generated - which is not the case.

**Comment by dimitrisnl** [comment]
@LucianBuzzo 	Do we want to keep the asRendition wrapper?

**Comment by LucianBuzzo** [comment]
@dimitrisnl Yep we want to keep it (now and in the future) - for this PR, the functionality of the lib should not change at all

**Comment by dimitrisnl** [comment]
Some failing tests, yet there are no visible changes in Alert, Pill & Navbar. Trying to figure it out.

**Comment by dimitrisnl** [comment]
@resin-ci 	retest

File: src/common-types.d.ts [hunk 1]
```diff
@@ -125,6 +125,221 @@ interface Coloring {
...
+interface BoxProps extends DefaultProps, Tooltip {
+	flex?: string | string[];
+	order?: number | string | Array<number | string>;
+}
+
+/* Used by:
+	1. Txt.tsx
+	2. Pill.tsx
+*/
+interface TxtProps extends DefaultProps, Tooltip {
```
Comment (sradevski) [inline src/common-types.d.ts]: I am commenting on the `TxtProps`, but most of these "common types" are not actually necessary. What you need to do is export the interface from the component (for example, `export interface TxtProps` in `TxtProps`, and then import it from there in the `Pill` component in this case.

File: src/components/Txt.tsx [hunk 2]
```diff
@@ -1,4 +1,3 @@
-import { Theme, TxtProps } from 'rendition';
```
Comment (LucianBuzzo) [inline src/components/Txt.tsx]: This seems to be missing the prop interface declaration

Comment (LucianBuzzo) [reply src/components/Txt.tsx]: Is it inferred automatically?

File: typings/rendition.d.ts [hunk 3]
```diff
@@ -1,242 +1,12 @@
 declare module 'rendition' {
```
Comment (LucianBuzzo) [inline typings/rendition.d.ts]: Can we completely delete this file?

**Comment by LucianBuzzo** [comment]
When I tested this against the UI it seems that the "common-types" are not picked up automatically and may need to be explicitly imported

File: src/common-types.d.ts (hunk 1 above)
Comment (dimitrisnl) [reply src/common-types.d.ts]: makes sense @sradevski Thanks for the pointer

Comment (dimitrisnl) [reply src/common-types.d.ts]: I tried it and it looks good, but my problem lies with the 'extra.d.ts' and 'unstable.d.ts' files, where I can't reference these typings relatevely.

I might have to break those apart.

File: typings/rendition.d.ts (hunk 3 above)
Comment (dimitrisnl) [reply typings/rendition.d.ts]: We should, left it for now in order to tackle the 'Form' typings a bit later.

**Comment by dimitrisnl** [comment]
@LucianBuzzo 

Isn't this in direct conflict with having a `rendition.d.ts` file? 
`common-types` is essentially the same but renamed.

Now it gets a little weird, with having the dashboard look for typings this way

```
import { Coloring, RenderableElementWithProps, Theme } from 'rendition';
```

**Comment by dimitrisnl** [comment]
@LucianBuzzo 
Screen Shot 2019-04-04 at 6 29 25 PM

The same thing happens when explicitly importing the types. Notice the error at the top right of the image.

The `.ts` file has no errors during compilation which is baffling

**Comment by LucianBuzzo** [comment]
@dimitrisnl The import should have a path I think, like:
```
import { Colouring } from '../common-types'
```

**Comment by dimitrisnl** [comment]
@LucianBuzzo That would be ideal if the other modules (extra.d.ts, unstable.d.ts) could handle relative paths.

I figured it out, tsconfing wouldn't pick these typings before.

File: src/components/Filters/index.tsx [hunk 4]
```diff
@@ -86,6 +132,20 @@ interface FiltersState {
 	schema: JSONSchema6;
 }
 
+interface FiltersProps extends DefaultProps {
```
Comment (sradevski) [inline src/components/Filters/index.tsx]: Maybe this needs to be exported as well?

File: src/components/Pill.tsx [hunk 5]
```diff
@@ -1,10 +1,17 @@
...
 import * as React from 'react';
 import { compose } from 'recompose';
-import { PillProps } from 'rendition';
 import styled, { withTheme } from 'styled-components';
 import asRendition from '../asRendition';
 import { getColor, withProps } from '../utils';
-import Txt from './Txt';
+import Txt, { TxtProps } from './Txt';
+
+export type PillProps = Coloring &
```
Comment (sradevski) [inline src/components/Pill.tsx]: Maybe make it an interface for consistency purposes.

File: typings/common-types.d.ts [hunk 6]
```diff
@@ -0,0 +1,167 @@
+declare module 'common-types' {
```
Comment (LucianBuzzo) [inline typings/common-types.d.ts]: Why is this a module instead of a simple `.ts` file that you can import interfaces from?

Comment (dimitrisnl) [reply typings/common-types.d.ts]: No real reason. I assumed that it made sense for types to be in a .ts' file. Should I update it?

Note that you can't have relative paths in declaration files, so if I wanted to reference something in `unstable.d.ts` I had to have `common-types` the same format.

**Comment by dimitrisnl** [comment]
@LucianBuzzo I followed up with dropping the '.ts' files in favour of simple typescript ones.

**Comment by dimitrisnl** [comment]
Super interesting, some snapshots changed - investigating.

**Pull request merged**